compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -991,73 +1020,66 @@ private STNode parseFuncDefinition(STNode metadata, boolean isObjectMember, STNo
      * function-type-descriptor := `function` function-signature
      * </code>
      *
-     * @param context Parsing context
      * @param metadata Preceding metadata
+     * @param visibilityQualifier
      * @param qualifiers Preceding visibility qualifier
      * @param isObjectMember Whether object member or not
      * @param isObjectTypeDesc Whether object type or not
      * @return Parsed node
      */
-    private STNode parseFuncDefOrFuncTypeDesc(ParserRuleContext context, STNode metadata, List<STNode> qualifiers,
+    private STNode parseFuncDefOrFuncTypeDesc(STNode metadata, STNode visibilityQualifier, List<STNode> qualifiers,
```

Comment:
We already did  this change when implementing new service changes in `service-typing` branch. There also had to move validation to parsing end. No null checks at the middle. Only at the end validation. Here in master we have null checks in the middle since I didn't align parseResource. When syncing they will be removed. 

Anyway if we treat it as a single list,  at the end validation we will have to extract it from the list before validating. WDYT?

Reply:
ack.
Will it handle duplicate public qualifiers?

Reply:
- As of now public qualifier is not a part of parseToplvelQualifiers. 
- It is parsed only one time and in case duplicated will be removed under `TOP_LEVEL_NODE_WITHOUT_MODIFIER` at recovery. 

I think that would be fine. It is very unlikely someone would insert something like `public public student;` Although they did, don't feel like we have to give a comprehensive error msg.

Reply:
ok